### ISSUE TITLE
Proposing Vert.x-Compliant client with arango-db

### ISSUE THREAD

**Issue by kostya05983** [reactiverse/reactiverse#7]
## Project Name
vertx-arangodb-client or vertx-arango-client

package can be like io.reactiverse:vertx-arango-client:1.0.0

## Project Goal
Arangodb is a fast multi-layer database. More information you can find [here](https://github.com/arangodb/arangodb)

I want to provide better compatibility of vert.x with arangodb and use arangodb with code-gen such as rxjava2 and so on.

## Licensing 
Apache License 2.0

## Motivation to the join the reactiverse
Develop such module together more funny and I don't want to make mistakes in architecture, which I won't be able to fix in future, I think, together we can develop more reliable library.

## Existing implementation
I made an a common interface with async handlers from vert.x like as in vertx-mongo-client.
Also I have a travis CI, which now I'm trying to set up.
Questions in architecture, which I try to solve:
* what is better to have a collection interface or an database interface where we put in methods the collection name?
* And what is better to use common jsonObject or just BaseDocument from arango or provide both variants?
Code can be found [here](https://github.com/kostya05983/vertx-arangodb)

What do you think?

**Comment by vietj** [comment]
I think we can discuss implementation details later, for now we should settle first on the project :)

**Comment by kostya05983** [comment]
@vietj, ok I deleted questions, which we can discuss later. I think, it's ready to vote. Also, @jponge please see it.

**Comment by jponge** [comment]
Looks interesting to me. Is there a prototype available?

(note: do not expect any further interaction from me for the next 2 weeks)

**Comment by kostya05983** [comment]
Yes, I add an example in tests of how to use it with testContainers. You can see it [here](https://github.com/kostya05983/vertx-arangodb/blob/master/src/test/java/io/vertx/ext/arango/ArangoGetDocumentTest.java). 

Next I want to provide configuration of api with familiar vert.x's jsonObject.

**Comment by vietj** [comment]
thanks @kostya05983 the users of the reactiverse org will deal with this, note that some of them are on PTO and it might take some time to process

**Comment by vietj** [comment]
I think we can start a vote next week

**Comment by cescoffier** [comment]
Looks good to me. I never heard about Arango, so now... I'm curious ;-)

**Comment by tsegismont** [comment]
So Arango is like Azure's Cosmos, right? As @cescoffier I had not heard about it before, but it looks interesting. Are you using this in production with Vert.x ?

**Comment by kostya05983** [comment]
Thanks, @tsegismont, No, I'm not using it in production. The one of key feature that arango is multi-model. And it can be useful if someone wants to store graphs in vert.x stack or store something else. Also, I had not heard about Azure's Cosmos before :)

**Comment by tsegismont** [comment]
Ok. I asked about your usage of Arango to understand your motivation. We
expect modules in reactiverse to keep up with Vert.x releases so that's
some work.

**Comment by kostya05983** [comment]
@tsegismont , More information about where this driver can be useful and my path to this idea. I'm worked in company, where we were using mongoClient. And when we made a lot of writes with updates it was really slow. After that I found an arangodb and I tried it in one competition, but it wasn't so comfortably to use, because it doesn't have client such as like vertx-mongo-client with compatible methods for vertx, that's why I decided to write such client.
[Image with benchmarks here](https://www.arangodb.com/wp-content/uploads/2018/02/UPDATE-Benchmark-2018.001.jpeg)

**Comment by kostya05983** [comment]
Helo, @vietj what's next? When will the vote over?

**Comment by jponge** [comment]
This is definitely interesting, but projects that come to _reactiverse_ should already have an implementation. So I suggest that you first implement it in your repository, then seek to contribute here.

**Comment by pmlopes** [comment]
Hi @kostya05983 I think an arangodb client vert.x friendly is a good addition to the ecosystem. I've spent some time looking at your repo and before I give my thumbs up, I'd like to see that the code would be a bit more mature than a PoC. What I mean is, I understand that the code follows the vert.x style of client interface + implementation and you say you'd like to have rx support. In order to do this the code misses the required annotations and the `package-info.java` file to enable it.

I think if some more effort would be put on it (so it could be used from the polyglot side) then I think everyone would be more comfortable to give their vote.

**Comment by kostya05983** [comment]
Thanks @pmlopes. I will try to make it more polyglot.
